### PR TITLE
feat(social-media-share): serve different static index.html files based on path

### LINKED ISSUE #361: Optimise agency pages for Search Engine Optimisation (SEO)
## Problem

The site currently ranks very low on Google search. This means that the visibility of the site is severely compromised and many citizens are likely to try looking up the questions they have on Google.

## Solution

Optimise agency pages and question pages to improve Google search rankings.

(Details on the specifics of the solution(s) will be added as more research as done.)

### LINKED ISSUE #403: Explore react rendering options for better SEO
## Problem 

Search crawlers need HTML to understand what the page is doing. However, as we are using React to build the page, the crawler will first see Javascript code, which it then has to send to the renderer to render recursively until either 1) site is fully rendered or 2) crawl budget is exceeded. The latter may mean that some of our pages are not indexed and thus not searchable.

## Solution

Explore rendering options other than the default client-side rendering (which I assume we are using). These options include:

- Client-side rendering with bootstrapped data
- Server-side rendering to static content
- Server-side rendering with rehydration
- Pre-rendering to static content
- Pre-rendering with rehydration

There is also isomorphic react, where we can consider using Next.js.

Helpful link: https://www.toptal.com/react/react-seo-best-practices

### PR DESCRIPTION
**Pull request by LinHuiqing**
## Problem

Social media crawlers are only able to see initial files served statically by server (i.e. `index.html`). To make our website more shareable on social media, these initial files should be populated with titles, meta description and open graph tags.

Related to #394 

## Solution

Get `server/src/bootstrap/index.ts` to modify `index.html` before serving them.

**Features**:

- (server/src/bootstrap/index.ts) send modified index.html for /agency/:shortname and /questions/:id paths for production

**Improvements**:

- (client/public/index.html) added og tags

## Before & After Screenshots

**BEFORE**:

Social media link preview:

Screenshot 2021-10-05 at 5 07 44 PM

**AFTER**:

Main AskGov page:

Screenshot 2021-10-05 at 5 13 29 PM

Agency-specific page:

Screenshot 2021-10-05 at 5 13 56 PM

Post page:

Screenshot 2021-10-05 at 5 14 58 PM

## Tests

Use [Facebook Sharing Debugger](https://developers.facebook.com/tools/debug/) to check that the title, meta description and og tags are populated. Specifically, test this for the following pages:

1. Main AskGov page
2. Agency-specific page
3. Post page

## Deploy Notes

**New dependencies**:

- `dependency` on `server`: 
  - `sanitize-html`
  - `cheerio`

**New dev dependencies**:

- `dependency` on `server`:  `@types/sanitize-html`

### CODE REVIEW

**Comment by LinHuiqing** [comment]
To be honest, `og:url` is still missing for some reason :/ not quite sure why or how to debug that?

File: server/src/bootstrap/index.ts [hunk 1]
```diff
@@ -194,9 +198,100 @@ if (baseConfig.nodeEnv === Environment.Prod) {
...
+          'content',
+          `Answers from ${req.params.shortname.toUpperCase()}`,
+        )
+        $('meta[property="og:description"]').attr(
+          'content',
+          `Answers from ${req.params.shortname.toUpperCase()}`,
+        )
+      }
+      res.header('content-type', 'text/html').send($.html())
+    })()
```
Comment (LoneRifle) [inline server/src/bootstrap/index.ts]: Why wrap it up like this? You can simply make the `(req, res) => { ... }` handler async

File: server/src/bootstrap/index.ts [hunk 2]
```diff
@@ -194,9 +198,100 @@ if (baseConfig.nodeEnv === Environment.Prod) {
...
+  })
+
+  app.get('*', (req, res) => {
+    const $ = cheerio.load(index)
+    $('meta[property="og:url"]').attr(
+      'content',
+      `${req.protocol}://${req.hostname}${req.originalUrl.toLowerCase()}`,
+    )
+    res.header('content-type', 'text/html').send($.html())
+  })
```
Comment (LoneRifle) [inline server/src/bootstrap/index.ts]: I think this is slightly spurious, so maybe we can drop this =)

**Review by LoneRifle — CHANGES_REQUESTED** [review]
Final thoughts:
- At this point, the implementation is fairly involved. Per earlier comments I made the other time, extract this out into its own module, following the structure as you find in `modules/`. So perhaps you could define something like `modules/web` and go from there with controller, route, service etc
- What do you do when the question or agency does not exist?

File: server/src/bootstrap/index.ts (hunk 2 above)
Comment (LinHuiqing) [reply server/src/bootstrap/index.ts]: okays

**Comment by LinHuiqing** [comment]
> * What do you do when the question or agency does not exist?

For this, what should we see actually? Should it be the default or some kind of "not found" page? Currently not found pages give the default preview too :/ 

Tangent: https://staging.ask.gov.sg/agency/no gives a blank page rather than a not found page. This is a bug, right?

**Comment by LoneRifle** [comment]
It's a missing feature, to be more precise. I would think that you could basically serve the not found page when you can't find the agency or question, and return http status code 404, thereby also solving #256

File: server/src/routes/index.ts [hunk 3]
```diff
@@ -58,5 +61,6 @@ export const api = (options: ApiRouterOptions): express.Router => {
   router.use('/posts/answers', routeAnswers(options.answers))
   router.use('/agencies', routeAgencies({ controller: options.agency }))
   router.use('/enquiries', routeEnquiries({ controller: options.enquiries }))
+  router.use('/web', routeWeb({ controller: options.web }))
```
Comment (LoneRifle) [inline server/src/routes/index.ts]: This isn't an api route..

**Review by LoneRifle — APPROVED** [review]

**Pull request merged**